File: 110. Balanced Binary Tree/110. Balanced Binary Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int checkBalance(struct TreeNode* root, bool* balanced) {
    if (!root) return 0;
    
    int leftHeight = checkBalance(root->left, balanced);
    int rightHeight = checkBalance(root->right, balanced);
    
    if (abs(leftHeight - rightHeight) > 1) *balanced = false;
    return 1 + (leftHeight > rightHeight ? leftHeight : rightHeight);
}

bool isBalanced(struct TreeNode* root) {
    bool balanced = true;
    checkBalance(root, &balanced);
    return balanced;
}

File: 110. Balanced Binary Tree/110. Balanced Binary Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    fn check_balance(root: Option<Rc<RefCell<TreeNode>>>) -> (bool, i32) {
        match root {
            None => (true, 0),
            Some(node) => {
                let node = node.borrow();
                let (left_balanced, left_height) = Self::check_balance(node.left.clone());
                let (right_balanced, right_height) = Self::check_balance(node.right.clone());
                
                let balanced = left_balanced && right_balanced && 
                             (left_height - right_height).abs() <= 1;
                (balanced, 1 + left_height.max(right_height))
            }
        }
    }
    
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::check_balance(root).0
    }
}